### DISCUSSION TITLE
Random number generation

### DISCUSSION THREAD

**Discussion by jornlp** [contiki-ng/contiki-ng#2816] [General]
Hi, 

Thank you for the nice work on the Contiki-ng OS.

While studying some work on random number generation I noticed the Blackhat presentation of[ Bishop Fox](https://bishopfox.com/blog/youre-doing-iot-rng) from august last year, referring to Contiki-ng as not properly generating random numbers which could lead to security vulnerabilities. I was wondering if you were notified, and if any measures were taken.

kind regards,

Jorn

**Comment by nvt** [comment]
Yes, we were notified of the issue. Since then, a CSPRNG has been merged into Contiki-NG through PR #522, which had been in the queue even before the notification. There still remains some code that needs to be updated to use this CSPRNG, however.

**Reply by jornlp** [reply]
Thanks for the info. 

I might miss something, but I am wondering why [os/net/app-layer/coap/tinydtls-support/dtls-support.c](https://github.com/contiki-ng/contiki-ng/blob/3b535924a2e96607ce3443b518a6abaf6f6db034/os/net/app-layer/coap/tinydtls-support/dtls-support.c#L96) is using the random_rand() which uses the insecure rand().
(see [os/lib/random.c](https://github.com/contiki-ng/contiki-ng/blob/3b535924a2e96607ce3443b518a6abaf6f6db034/os/lib/random.c#L45), [arch/cpu/nrf52840/dev/random.c](https://github.com/contiki-ng/contiki-ng/blob/3b535924a2e96607ce3443b518a6abaf6f6db034/arch/cpu/nrf52840/dev/random.c#L51), [arch/cpu/gecko/os/random-arch.c](https://github.com/contiki-ng/contiki-ng/blob/3b535924a2e96607ce3443b518a6abaf6f6db034/arch/cpu/gecko/os/random-arch.c#L60))
It is also the case in the separate tinydtls repository.

In the ecc-ecdh.h example it also uses the insecure version inside ecc_set_random, which pave the way to insecure implementations by users of your OS.

with kind regards,

Jorn
